### PR TITLE
[Mono] Implement eager finalization of WeakReference

### PR DESCRIPTION
**Pull request by filipnavara**

### CODE REVIEW

File: src/mono/mono/metadata/gc.c [hunk 1]
```diff
@@ -249,6 +249,15 @@ mono_gc_run_finalize (void *obj, void *data)
...
 	}
 
+	if (o->vtable->klass == mono_defaults.weakreference_class ||
+	    o->vtable->klass == mono_defaults.generic_weakreference_class) {
+		MonoWeakReference *wr = (MonoWeakReference*)o;
+		MonoGCHandle gc_handle = (MonoGCHandle)(wr->handleAndKind & ~(gsize)1);
+		mono_gchandle_free_internal (gc_handle);
+		wr->handleAndKind &= (gsize)1;
+		return;
+	}
```
Comment (filipnavara) [inline src/mono/mono/metadata/gc.c:259]: Hmm... maybe I misunderstood when the `mono_gc_run_finalize` method is run and on what thread.

**Comment by naricc** [comment]
I'm not sure I understand what is happening here. What does making finalization eager actually mean for WeakRefs? What is the behavior supposed to be?

**Comment by VSadov** [comment]
WeakReference types are finalizable types. The weak references could be used in big quantities and that can have bad effects on finalization queue.   
The finalization of weak references is very trivial though - it is basically closing the handle. It is generally cheaper to just close the handle when GC finds unreachable weak reference, instead of posting it to the finalization queue. 

The eager approach also has some reliability advantages. Since eager finalization happens when managed threads are stopped, there is no chance of getting into races between finalization and concurrent use of the weak references from other threads.

CoreCLR had eager finalization of weak references for a long time. NativeAOT has added that recently - https://github.com/dotnet/runtime/pull/75436

Another advantage is that if Mono implements eager finalization, it would be able to use the shared implementation of  `WeakReference` and `WeakReference<T>` without any ifdefs.

File: src/mono/mono/metadata/class-internals.h [hunk 2]
```diff
@@ -921,6 +921,8 @@ typedef struct {
 	MonoClass *generic_ienumerator_class;
 	MonoClass *alc_class;
 	MonoClass *appcontext_class;
+	MonoClass *weakreference_class;
+	MonoClass *generic_weakreference_class;
```
Comment (lambdageek) [inline src/mono/mono/metadata/class-internals.h:925]: nit. we don't need this in `mono_defaults`.
in sgen-mono.c` write:

```c
#include "metadata/class-internals.h"

...

static GENERATE_TRY_GET_CLASS_WITH_CACHE_DECL(weakreference);
static
GENERATE_TRY_GET_CLASS_WITH_CACHE_DECL(generic_weakreference);

GENERATE_TRY_GET_CLASS_WITH_CACHE (weakreference, "System", "WeakReference");
GENERATE_TRY_GET_CLASS_WITH_CACHE (generic_weakreference, "System", "WeakReference`1")
```

and then call `mono_class_try_get_weakreference_class` and `mono_class_try_get_generic_weakreference_class`

**Comment by filipnavara** [comment]
Thanks @VSadov for explaining it better than I would 👍

**Review by lambdageek — APPROVED** [review]
lgtm. although i'm not sure we should put weakreference into `mono_defaults`

File: src/mono/mono/metadata/class-internals.h (hunk 2 above)
Comment (filipnavara) [reply src/mono/mono/metadata/class-internals.h:925]: Makes sense, I will update it.

**Review by naricc — APPROVED** [review]

File: src/mono/mono/metadata/class-internals.h (hunk 2 above)
Comment (filipnavara) [reply src/mono/mono/metadata/class-internals.h:925]: Actually, on second thought...

```
	/*
	Not using GENERATE_TRY_GET_CLASS_WITH_CACHE_DECL as this type is heavily checked by sgen when computing finalization.
	*/
	mono_defaults.critical_finalizer_object = mono_class_try_load_from_name (mono_defaults.corlib,
```

This is accessed just as often in the same code path and it was intentionally not done through `GENERATE_TRY_GET_CLASS_WITH_CACHE_DECL`. I can push the change to use the `GENERATE_*` methods but I want to double check first.

Comment (lambdageek) [reply src/mono/mono/metadata/class-internals.h:925]: `sgen_client_object_has_critical_finalizer` is called from multiple files, so generating an accessor method (assuming it's not done in a .h file) would make it not inlineable.
on the other hand WeakReference is (so far) only looked up from one .c file, and I'm suggesting putting the call to `GENEREATE` in the same .c file.  So we should get inlined code.

I don't feel very strongly about this, however. either way is an improvement over the status quo, I think

Comment (BrzVlad) [reply src/mono/mono/metadata/class-internals.h:925]: I agree with @filipnavara that it wouldn't be safe to do class loading during a GC.

**Comment by BrzVlad** [comment]
This change could further be improved by avoiding a useless object allocation. In `sgen_finalize_in_range`, `copy_func` is used to promote / mark the object, since it still needs to be alive to be passed as `this` to the finalizer. This should no longer be needed if it's eagerly finalized.

**Comment by VSadov** [comment]
@BrzVlad As I understand the suggestion, we need to move the specialcasing of weak references a bit earlier - from `sgen_queue_finalization_entry` up into `sgen_finalize_in_range`. Basically instad of `copy_func (&copy, queue);` we can just `SGEN_HASH_TABLE_FOREACH_REMOVE (TRUE);` and continue.

Is this the idea?

**Comment by BrzVlad** [comment]
@VSadov Yes, that is correct

**Comment by filipnavara** [comment]
Thanks for taking care of the last feedback, @VSadov!

**Comment by VSadov** [comment]
I assume the change addresses the suggestions and once `llvmfullaot` finishes we can merge the change.

**Review by BrzVlad — APPROVED** [review]

**Comment by uweigand** [comment]
As of this commit, I'm now seeing failures when running the libs.tests suite on s390x (Mono-based runtime):
```
  Process terminated. Assertion failed.
  WeakReference<T> finalizer should never run
```
Any suggestions what this could be, or how to further debug?

**Comment by BrzVlad** [comment]
I remember seeing this assertion locally but I didn't get to investigate it. I'll see if I can repro this week

**Comment by filipnavara** [comment]
@uweigand Do you have any details about which test is failing?

**Comment by uweigand** [comment]
> @uweigand Do you have any details about which test is failing?

It varies, I've seen multiple different tests failing.  One that seems to fail most of the time is `System.IO.FileSystem.Watcher.Tests`.

**Comment by uweigand** [comment]
Also interesting is that for some unknown reason, I'm not seeing any of these failures in the CI here:
https://dev.azure.com/dnceng-public/public/_build?definitionId=148
(which is doing a cross-build followed by native execution), but I'm seeing the failures all the time when building + testing fully natively on my own s390x machine.

**Comment by BrzVlad** [comment]
> > @uweigand Do you have any details about which test is failing?
> 
> It varies, I've seen multiple different tests failing. One that seems to fail most of the time is `System.IO.FileSystem.Watcher.Tests`.

I can reproduce, will take a look later

**Comment by uweigand** [comment]
> > > @uweigand Do you have any details about which test is failing?
> > 
> > 
> > It varies, I've seen multiple different tests failing. One that seems to fail most of the time is `System.IO.FileSystem.Watcher.Tests`.
> 
> I can reproduce, will take a look later

@BrzVlad did you have a chance to look into this?

**Comment by BrzVlad** [comment]
@uweigand Should be fixed by https://github.com/dotnet/runtime/pull/77170

**Comment by filipnavara** [comment]
Thanks a lot!

**Comment by uweigand** [comment]
Yes, this has fixed the test suite on s390x.   Thanks!

**Pull request merged**